test/core/pipeline/features/test_run.py

```diff
@@ -5392,3 +5392,131 @@ def run(self, dataframe: pd.DataFrame) -> Dict[str, Any]:
             )
         ],
     )
+
+
+@given(
+    "a pipeline where a single component connects multiple sockets to the same receiver socket",
+    target_fixture="pipeline_data",
+)
+def pipeline_single_component_many_sockets_same_target(pipeline_class):
+    joiner = BranchJoiner(type_=str)
+
+    routes = [
+        {"condition": "{{query == 'route_1'}}", "output": "{{query}}", "output_name": "route_1", "output_type": str},
+        {"condition": "{{query == 'route_2'}}", "output": "{{query}}", "output_name": "route_2", "output_type": str},
+    ]
+
+    router = ConditionalRouter(routes=routes)
+
+    pp = pipeline_class(max_runs_per_component=1)
+
+    pp.add_component("joiner", joiner)
+    pp.add_component("router", router)
+
+    pp.connect("router.route_1", "joiner.value")
+    pp.connect("router.route_2", "joiner.value")
+
+    return (
+        pp,
+        [
+            PipelineRunData(
+                inputs={"router": {"query": "route_1"}},
+                expected_outputs={"joiner": {"value": "route_1"}},
+                expected_component_calls={("router", 1): {"query": "route_1"}, ("joiner", 1): {"value": ["route_1"]}},
+            )
+        ],
+    )
+
+
+@given(
+    "a pipeline where a component in a cycle provides inputs for a component outside the cycle in one iteration and no input in another iteration",
+    target_fixture="pipeline_data",
+)
+def pipeline_component_cycle_input_no_input(pipeline_class):
+    joiner = BranchJoiner(type_=str)
+
+    routes = [
+        {
+            "condition": "{{query == 'iterationiterationiterationiteration'}}",
+            "output": "{{query}}",
+            "output_name": "exit",
+            "output_type": str,
+        },
+        {
+            "condition": "{{query != 'iterationiterationiterationiteration'}}",
+            "output": "{{query}}",
+            "output_name": "continue",
+            "output_type": str,
+        },
+    ]
+
+    template = "{{query ~ query}}"
+
+    builder = PromptBuilder(template=template)
+
+    router = ConditionalRouter(routes=routes)
+
+    outside_builder = PromptBuilder(
+        template="{{cycle_output ~ delayed_input}}", required_variables=["cycle_output", "delayed_input"]
+    )
+
+    outside_routes = [
+        {
+            "condition": "{{query == 'iterationiteration'}}",
+            "output": "{{query}}",
+            "output_name": "cycle_output",
+            "output_type": str,
+        },
+        {
+            "condition": "{{query != 'iterationiteration'}}",
+            "output": "{{query}}",
+            "output_name": "no_output",
+            "output_type": str,
+        },
+    ]
+
+    outside_router = ConditionalRouter(routes=outside_routes)
+
+    pp = pipeline_class(max_runs_per_component=1)
```

Comment:
Why does `max_runs_per_component=1` work here if we test cycles? I expected let's say 10 instead. 2 should work too but 1 I would expect to fail.

Reply:
Good point! The limit was added by accident, I just copied that part from somewhere where I set a limit.

Why does it work though? 
It's something that I wanted to bring up earlier but I forgot about it. We actually have (always had) and off-by-one error in our max_runs_per_component. We start visits from zero but we count the limit by checking `if visits > max_runs_per_component`.

We could change that but it would potentially be a breaking change for some people.

Reply:
Got it. Thanks!
 off-by-one error... of course. :D